File: backend/src/main/java/com/styp/cenate/service/ipress/IpressService.java
package com.styp.cenate.service.ipress;

import com.styp.cenate.dto.IpressRequest;
import com.styp.cenate.dto.IpressResponse;
import com.styp.cenate.dto.ActualizarModalidadIpressRequest;
import com.styp.cenate.dto.ModuloDisponibleDTO;
import java.util.List;

/**
 * Servicio para gestionar las IPRESS (Instituciones Prestadoras de Servicios de Salud)
 */
public interface IpressService {

    // ===========================
    // CONSULTAS (READ)
    // ===========================
    List<IpressResponse> getAllIpress();

    List<IpressResponse> getIpressActivas();

    IpressResponse getIpressById(Long id);

    List<IpressResponse> searchIpress(String searchTerm);

    List<IpressResponse> getIpressActivasPorRed(Long idRed);

    // ===========================
    // CREAR (CREATE)
    // ===========================
    /**
     * Crea una nueva IPRESS
     * @param request datos de la IPRESS a crear
     * @return IPRESS creada
     */
    IpressResponse createIpress(IpressRequest request);

    // ===========================
    // ACTUALIZAR (UPDATE)
    // ===========================
    /**
     * Actualiza una IPRESS existente
     * @param id ID de la IPRESS a actualizar
     * @param request datos actualizados
     * @return IPRESS actualizada
     */
    IpressResponse updateIpress(Long id, IpressRequest request);

    /**
     * Actualiza la modalidad de atención de la IPRESS del usuario logueado (Personal Externo)
     * @param request datos de modalidad a actualizar
     * @return IPRESS actualizada
     */
    IpressResponse actualizarModalidadPorUsuarioActual(ActualizarModalidadIpressRequest request);

    /**
     * Obtiene la IPRESS asignada al usuario logueado (Personal Externo)
     * @return IPRESS del usuario logueado
     */
    IpressResponse obtenerIpressPorUsuarioActual();

    /**
     * Obtiene los módulos disponibles (habilitados) para la IPRESS del usuario logueado
     * @return Lista de módulos disponibles con información de navegación
     */
    List<ModuloDisponibleDTO> obtenerModulosDisponibles();

    /**
     * Obtiene los módulos disponibles para una IPRESS específica
     * @param idIpress ID de la IPRESS
     * @return Lista de módulos disponibles
     */
    List<ModuloDisponibleDTO> obtenerModulosDisponiblesPorIpress(Long idIpress);

    // ===========================
    // ELIMINAR (DELETE)
    // ===========================
    /**
     * Elimina una IPRESS (solo SUPERADMIN)
     * @param id ID de la IPRESS a eliminar
     */
    void deleteIpress(Long id);
}

File: frontend/src/services/ipressService.js
// ========================================================================
// 🏥 ipressService.js – Servicio para IPRESS (CENATE 2026)
// ------------------------------------------------------------------------
// Servicio para gestionar Instituciones Prestadoras de Servicios de Salud
// ========================================================================

import { apiClient } from "../lib/apiClient";

class IpressService {
  /**
   * Obtener todas las IPRESS
   * @returns {Promise<Array>} Lista de IPRESS
   */
  async obtenerTodas() {
    try {
      const data = await apiClient.get("/ipress", true);
      return Array.isArray(data) ? data : [];
    } catch (error) {
      console.error("❌ Error al obtener IPRESS:", error);
      throw error;
    }
  }

  /**
   * Obtener IPRESS activas
   * @returns {Promise<Array>} Lista de IPRESS activas
   */
  async obtenerActivas() {
    try {
      const data = await apiClient.get("/ipress/activas", true);
      return Array.isArray(data) ? data : [];
    } catch (error) {
      console.error("❌ Error al obtener IPRESS activas:", error);
      throw error;
    }
  }

  /**
   * Obtener una IPRESS por ID
   * @param {number} id - ID de la IPRESS
   * @returns {Promise<Object>} IPRESS encontrada
   */
  async obtenerPorId(id) {
    try {
      const data = await apiClient.get(`/ipress/${id}`, true);
      return data;
    } catch (error) {
      console.error(`❌ Error al obtener IPRESS ${id}:`, error);
      throw error;
    }
  }

  /**
   * Buscar IPRESS por nombre
   * @param {string} termino - Término de búsqueda
   * @returns {Promise<Array>} Lista de IPRESS encontradas
   */
  async buscar(termino) {
    try {
      const data = await apiClient.get(`/ipress/buscar?q=${encodeURIComponent(termino)}`, true);
      return Array.isArray(data) ? data : [];
    } catch (error) {
      console.error(`❌ Error al buscar IPRESS con término "${termino}":`, error);
      throw error;
    }
  }

  /**
   * Crear una nueva IPRESS
   * @param {Object} ipressData - Datos de la IPRESS
   * @returns {Promise<Object>} IPRESS creada
   */
  async crear(ipressData) {
    try {
      const response = await apiClient.post("/ipress", ipressData, true);
      console.log("✅ IPRESS creada exitosamente");
      return response.data || response;
    } catch (error) {
      console.error("❌ Error al crear IPRESS:", error);
      throw error;
    }
  }

  /**
   * Actualizar una IPRESS existente
   * @param {number} id - ID de la IPRESS
   * @param {Object} ipressData - Datos actualizados
   * @returns {Promise<Object>} IPRESS actualizada
   */
  async actualizar(id, ipressData) {
    try {
      const response = await apiClient.put(`/ipress/${id}`, ipressData, true);
      console.log(`✅ IPRESS ${id} actualizada exitosamente`);
      return response.data || response;
    } catch (error) {
      console.error(`❌ Error al actualizar IPRESS ${id}:`, error);
      throw error;
    }
  }

  /**
   * Eliminar una IPRESS (solo SUPERADMIN)
   * @param {number} id - ID de la IPRESS
   * @returns {Promise<void>}
   */
  async eliminar(id) {
    try {
      await apiClient.delete(`/ipress/${id}`, true);
      console.log(`✅ IPRESS ${id} eliminada exitosamente`);
    } catch (error) {
      console.error(`❌ Error al eliminar IPRESS ${id}:`, error);
      throw error;
    }
  }

  /**
   * Obtener modalidades de atención activas
   * @returns {Promise<Array>} Lista de modalidades
   */
  async obtenerModalidadesActivas() {
    try {
      const data = await apiClient.get("/modalidades-atencion/activas", false);
      return Array.isArray(data) ? data : [];
    } catch (error) {
      console.error("❌ Error al obtener modalidades de atención:", error);
      return [];
    }
  }

  /**
   * Obtener la IPRESS asignada al usuario logueado (Personal Externo)
   * @returns {Promise<Object>} IPRESS del usuario
   */
  async obtenerMiIpress() {
    try {
      const response = await apiClient.get("/ipress/mi-ipress", true);
      return response.data || response;
    } catch (error) {
      console.error("❌ Error al obtener mi IPRESS:", error);
      throw error;
    }
  }

  /**
   * Actualizar modalidad de atención de la IPRESS del usuario logueado
   * @param {Object} modalidadData - Datos de modalidad a actualizar
   * @returns {Promise<Object>} IPRESS actualizada
   */
  async actualizarMiModalidad(modalidadData) {
    try {
      const response = await apiClient.patch("/ipress/mi-modalidad", modalidadData, true);
      console.log("✅ Modalidad de atención actualizada exitosamente");
      return response.data || response;
    } catch (error) {
      console.error("❌ Error al actualizar modalidad de atención:", error);
      throw error;
    }
  }
}

export default new IpressService();
